sympy/polys/rings.py

```diff
@@ -3006,3 +3206,252 @@ def sqf_list(f, all=False):
 
     def factor_list(f):
         return f.ring.dmp_factor_list(f)
+
+def monomial_extract(polynomials):
+    """
+    Extracts any common monomial from the polynomials.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import monomial_extract
+    >>> from sympy import ZZ, ring
+    >>> R, x, y = ring("x, y", ZZ)
+
+    >>> f = x + y
+    >>> g = x**2*y + x*y
+    >>> polynomials = [f, g]
+    >>> monomial_extract(polynomials)
+    ([x + y, x**2*y + x*y], 1)
+
+    >>> f = x**2*y
+    >>> g = x**2*y + x*y
+    >>> polynomials = [f, g]
+    >>> monomial_extract(polynomials)
+    ([x, x + 1], x*y)
+
+    """
+    ring = polynomials[0].ring
+    domain = ring.domain
+
+    zero_monom = ring.zero_monom
+    monoms = chain(*polynomials)
+
+    # Check for the presence of zero_monom in any polynomial
+    if any(zero_monom in poly for poly in polynomials):
+        return polynomials, domain.one
+
+    monom_gcd = tuple(map(min, zip(*monoms)))
+
+    if monom_gcd == ring.zero_monom:
+        return polynomials, domain.one
+    else:
+        d = ring({monom_gcd: ring.domain.one})
+        p = [pi.exquo(d) for pi in polynomials]  # TODO: Use monomial_ldiv
+        return p, d
+
+def _gcd_preprocess_polys(polynomials):
+    """
+    Simplify a list of polynomials whose gcd is wanted. Returns a possibly
+    longer list of simpler polynomials having the same gcd as the input.
+
+    This is done by eliminating symbols that can not be part of the gcd
+    because they do not appear in each item of the input. In the output
+    list, all items have exactly the same symbols. The set of those symbols
+    is also returned.
+
+    Examples
+    ========
+
+    >>> from sympy.polys.rings import _gcd_preprocess_polys
+    >>> from sympy import ring, ZZ
+    >>> R, x, y, z = ring("x, y, z", ZZ)
+
+    >>> f = x**2 + y**2
+    >>> g = x**2 - z**2
+    >>> h = y**3 + z**3
+    >>> polynomials = [f, g, h]
+    >>> _gcd_preprocess_polys(polynomials)
+    ([1], None)
```

Comment:
Showing only trivial doctest examples does not help the reader to understand what the method does.